### PR TITLE
Query number of upstream messages from Storage Integration

### PR DESCRIPTION
**Pull request by neoaggelos**
#### Summary

References https://github.com/TheThingsIndustries/lorawan-stack/pull/2873

This PR contains the API and CLI changes.

#### Changes

- Add `GetStoredApplicationUpCount` RPC API.
- Implement `ttn-lw-cli app storage count [application-id]` command.
- Implement `ttn-lw-cli dev storage count [application-id] [device-id]` command.

#### Testing

Locally

##### Regressions

None

#### Notes for Reviewers

This is already reviewed in referenced PR.

#### Checklist

- [X] Scope: The referenced issue is addressed, there are no unrelated changes.
- [X] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [X] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: api/applicationserver_integrations_storage.proto [hunk 1]
```diff
@@ -82,6 +80,43 @@ message GetStoredApplicationUpRequest {
...
+  // Count uplinks on a specific FPort only.
+  google.protobuf.UInt32Value f_port = 6;
+
+  // Count upstream messages that have arrived in the last minutes or hours. Cannot be used in conjunction with after and before.
+  google.protobuf.Duration last = 7;
+}
+
+
+message GetStoredApplicationUpCountResponse {
+  map<string,uint32> count = 1;
```
Comment (johanstokking) [inline api/applicationserver_integrations_storage.proto:118]: The map key here is the type, right? Please document that here

**Review by johanstokking — COMMENTED** [review]
See also https://github.com/TheThingsIndustries/lorawan-stack/issues/2509#issuecomment-900087387

File: api/applicationserver_integrations_storage.proto (hunk 1 above)
Comment (neoaggelos) [reply api/applicationserver_integrations_storage.proto:118]: This map key is the device ID, so the output looks like:

```json
{"count": {"dev1": 100, "dev2": 200}}
```

All requests are scoped to a single application.

**Review by htdvisser — COMMENTED** [review]
I'm missing CHANGELOG here for the new API and new CLI command.

**Review by adriansmares — APPROVED** [review]

File: api/applicationserver_integrations_storage.proto (hunk 1 above)
Comment (KrishnaIyer) [reply api/applicationserver_integrations_storage.proto:118]: We could actually echo the type in the response to help out the caller but we can add that later as well.

File: api/applicationserver_integrations_storage.proto [hunk 2]
```diff
@@ -82,6 +80,44 @@ message GetStoredApplicationUpRequest {
   google.protobuf.Duration last = 10;
 }
 
+message GetStoredApplicationUpCountRequest {
+  // Count upstream messages from all end devices of an application. Cannot be used in conjunction with end_device_ids.
+  ApplicationIdentifiers application_ids = 1;
+  // Count upstream messages from a single end device. Cannot be used in conjunction with application_ids.
+  EndDeviceIdentifiers end_device_ids = 2;
```
Comment (KrishnaIyer) [inline api/applicationserver_integrations_storage.proto:87]: We could use the `EntityIdentifiers` and assert the type to one of these two right?

File: cmd/ttn-lw-cli/commands/storage_integration.go [hunk 3]
```diff
@@ -114,15 +139,46 @@ var (
 			return getStoredUp(cmd, args, client, os.Stdout)
 		},
 	}
+	applicationsStorageCountCommand = &cobra.Command{
+		Use:   "count [application-id]",
+		Short: "Retrieve stored upstream messages",
```
Comment (KrishnaIyer) [inline cmd/ttn-lw-cli/commands/storage_integration.go]: I find this a bit confusing (and I guess so will users).
```
count [application-id] [device-id]
```
just returns a count where as

```
count [application-id]
```
Retrieves the actual messages.

**Review by KrishnaIyer — COMMENTED** [review]
I've a few question on the CLI but the rest looks good. Should we maybe wrap the "Unimplemented" error and present it nicely when someone makes this request with an OS backend?

File: api/applicationserver_integrations_storage.proto (hunk 2 above)
Comment (adriansmares) [reply api/applicationserver_integrations_storage.proto:87]: The HTTP bindings don't work with oneof fields, that's why these fields are explicit

File: cmd/ttn-lw-cli/commands/storage_integration.go (hunk 3 above)
Comment (adriansmares) [reply cmd/ttn-lw-cli/commands/storage_integration.go]: This is a copy paste failure I believe

Comment (neoaggelos) [reply cmd/ttn-lw-cli/commands/storage_integration.go]: Fixed

**Review by KrishnaIyer — APPROVED** [review]

**Pull request merged**